codebuild/builder.py

```diff
@@ -0,0 +1,731 @@
+        return [_replace_variables(e, variables) for e in value]
+
+    elif key_type == dict:
+        # Iterate each element and recursively apply the variables
+        return dict([(key, _replace_variables(value, variables)) for (key, value) in value.items()])
+
+    else:
+        # Unsupported, just return it
+        return value
+
+# Traverse the configurations to produce one for the specified
+def produce_config(build_spec, **additional_variables):
+
+    validate_build(build_spec)
+
+    # Build the list of config options to poll
+    configs = []
+
+    def process_config(map, key):
+        config = map[key]
+        configs.append(config)
+
+        config_arches = config.get('architectures')
+        if config_arches:
+            config_arch = config_arches.get(build_spec.arch)
+            if config_arch:
+                configs.append(config_arch)
+
+        return config
+
+
+    # Host isn't allowed to specify architectures
+    process_config(HOSTS, build_spec.host)
+    process_config(TARGETS, build_spec.target)
+
+    compiler = process_config(COMPILERS, build_spec.compiler)
+    process_config(compiler['versions'], build_spec.compiler_version)
+
+    new_version = {
+        'spec': build_spec,
+    }
+    # Iterate all keys and apply them
+    for key, default in KEYS.items():
+        new_version[key] = default
+
+        for config in configs:
+            override_key = '!' + key
+            if override_key in config:
+                # Handle overrides
+                new_version[key] = config[override_key]
+
+            elif key in config:
+                # By default, merge all values (except strings)
+                _apply_value(new_version, key, config[key])
+
+    # Default variables
+    replacements = {
+        'host': build_spec.host,
+        'compiler': build_spec.compiler,
+        'version': build_spec.compiler_version,
+        'target': build_spec.target,
+        'arch': build_spec.arch,
+        **additional_variables,
+    }
+
+    # Pull variables from the configs
+    for config in configs:
+        if 'variables' in config:
+            variables = config['variables']
+            assert type(variables) == dict
+
+            # Copy into the variables list
+            for k, v in variables.items():
+                replacements[k] = v
+
+    # Post process
+    new_version = _replace_variables(new_version, replacements)
+
+    return new_version
+
+########################################################################################################################
+# RUN BUILD
+########################################################################################################################
+
+def run_build(build_spec, is_dryrun):
+
+    if not is_dryrun:
+        import tempfile, shutil, subprocess
+
+    #TODO These platforms don't succeed when doing a RelWithDebInfo build
+    if build_spec.host in ("al2012", "manylinux"):
+        global BUILD_CONFIG
```

Comment:
Why not just cache it into a local and not mess with the global?

Reply:
This way when we fix these issues, I can just nuke this chunk and the problem goes away